[TensorExpr] CudaCodegen: restart counter for function names unique ID inside each codegen instantiation.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47255 [TensorExpr] Pick meaningful names for functions in TE codegen.
* **#47254 [TensorExpr] CudaCodegen: restart counter for function names unique ID inside each codegen instantiation.**
* #47253 [JIT] SubgraphUtils: add a function for generating a string name for a given graph.

CUDA codegen used a static global counter for picking names for
functions, but the functions only need to be unique in the scope of the
given codegen. This PR fixes that.

Differential Revision: [D24698271](https://our.internmc.facebook.com/intern/diff/D24698271)

Commit:
[TensorExpr] CudaCodegen: restart counter for function names unique ID inside each codegen instantiation.

CUDA codegen used a static global counter for picking names for
functions, but the functions only need to be unique in the scope of the
given codegen. This PR fixes that.

[ghstack-poisoned]

nickgg commented:
We have a `UniqueNameManager` for this purpose, why not reuse it?

bertmaher reviewed (approved):
Kinda fine with me whether to use this set or the UniqueNameManager so I'm accepting, but you're welcome to request changes if you feel strongly @nickgg

bertmaher commented:
Hmm, wait a second.  These names are supposed to be globally unique, aren't they?  Since we use a new codegen instance for each kernel, that property won't hold any more...

ZolotukhinM commented:
> Hmm, wait a second. These names are supposed to be globally unique, aren't they? Since we use a new codegen instance for each kernel, that property won't hold any more...

I'm not sure about that. From my reading of `CudaCodeGen::CompileToNVRTC` and `CudaCodeGen::call` the only requirement is that there are no duplicated symbols in the cuda program we generate, which holds true with this change as well. My understanding is that different instances of `CudaCodeGen` can not interfere with each other in any way since each of them would compile a separate "program". Is it not so?

ZolotukhinM commented:
> We have a `UniqueNameManager` for this purpose, why not reuse it?

Ah, I actually forgot about it, but it looks like `UniqueNameManager` currently only handles names for `Var`s and doesn't provide utilities to make strings unique. It is of course possible to add such API there, but this will be the only user of it for now. @nickgg would you prefer me to make such changes?

nickgg commented:
Oh nah, it's fine, if it's more trouble than it's worth then I wouldn't worry about it.

For the other thing, running fastRNN on GPU should make it clear whether or not the names need to be globally unique. @zheng-xq  may also know since he wrote the original.

Commit:
Update on "[TensorExpr] CudaCodegen: restart counter for function names unique ID inside each codegen instantiation."


CUDA codegen used a static global counter for picking names for
functions, but the functions only need to be unique in the scope of the
given codegen. This PR fixes that.

Differential Revision: [D24698271](https://our.internmc.facebook.com/intern/diff/D24698271)

[ghstack-poisoned]

ZolotukhinM commented:
> Oh nah, it's fine, if it's more trouble than it's worth then I wouldn't worry about it.
> 
> For the other thing, running fastRNN on GPU should make it clear whether or not the names need to be globally unique. @zheng-xq may also know since he wrote the original.

Just checked the FastRNNs, they seem to be running fine with this stack.

zheng-xq commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -658,12 +658,14 @@ class PrioritizeLoad : public IRMutator {
-  ++counter;
-  int64_t value = counter;
-  return func_prefix + "_" + std::to_string(value);
+  int64_t counter = 0;
```
I am fine with this change. But if the number of kernels is large, this could be quadratic. Can we move the counter to the class object scope as well, similar to taken_func_names? Thanks!

ZolotukhinM replied:
This will only be quadratic on number of kernels with the same body (executing the same operations), since the name is now reflecting operations we're doing in the body, so the collisions would actually be quite rare I think. Do you think we could really hit a scenario where we have a big number of identical kernels in the same program (note that in the current pipeline we usually have just one kernel in the program)?

nickgg replied:
I agree with you here that its unlikely to be a factor, but the fix in this case is pretty simple (change the set to a map to int64) it might be worth doing anyway.

Commit:
Update on "[TensorExpr] CudaCodegen: restart counter for function names unique ID inside each codegen instantiation."


CUDA codegen used a static global counter for picking names for
functions, but the functions only need to be unique in the scope of the
given codegen. This PR fixes that.

Differential Revision: [D24698271](https://our.internmc.facebook.com/intern/diff/D24698271)

[ghstack-poisoned]
